src/bytes/bytes_lemmas.gobra

```diff
@@ -0,0 +1,415 @@
+package bytes
+
+// +gobra
+
+import (
+	. "verification/utils/definitions"
+	sl "gobra-libs/byteslice"
+	bytes "gobra-libs/bytes"
+	. "bytes/spec"
+)
+
+decreases
+requires acc(sl.Bytes(b, 0, len(b)), _)
+pure func String(b []byte) string  {
+	return unfolding acc(sl.Bytes(b, 0, len(b)), _) in string(b)
+}
+
+// We cannot yet prove this with the current incomplete encoding of strings
+ghost
+requires 0 < p
+requires acc(sl.Bytes(a, 0, len(a)), p)
+requires acc(sl.Bytes(b, 0, len(b)), p)
+requires r == (String(a) == String(b))
+ensures  acc(sl.Bytes(a, 0, len(a)), p)
+ensures  acc(sl.Bytes(b, 0, len(b)), p)
+ensures  r == (sl.View(a) == sl.View(b))
+trusted
+decreases
+func stringEqualsImplViewEquals(r bool, a, b []byte, p perm)
+
+
+
+// Count
+
+ghost
+requires InRangeInc(idx, 0, len(s))
+requires forall j int :: { j in indices }{ s[j:j+len(sep)] } !(j in indices) ==> !(InRangeInc(j, 0, idx - len(sep))) || s[j:j+len(sep)] != sep || SetContainsInRange(indices, j-len(sep), j)
+requires forall j int :: {s[idx:][j:j+len(sep)]} InRangeInc(j, 0, len(s[idx:]) - len(sep)) ==> s[idx:][j:j+len(sep)] != sep
+ensures  forall j int :: { j in indices }{ s[j:j+len(sep)] } !(j in indices) ==> ( !InRangeInc(j, 0, len(s) - len(sep)) || s[j:j+len(sep)] != sep || SetContainsInRange(indices, j-len(sep), j))
+decreases
+func lemmaCountAux(s, sep seq[byte], indices set[int], idx int) {
+	assume forall j int :: {s[j:j+len(sep)]} InRangeInc(j, idx, len(s) - len(sep)) ==> s[j:j+len(sep)] != sep
+	assert forall j int :: { j in indices }{ s[j:j+len(sep)] } !(j in indices) && InRangeInc(j, 0, idx - len(sep)) && !SetContainsInRange(indices, j-len(sep), j) ==> s[j:j+len(sep)] != sep
+	assert forall j int :: { j in indices }{ s[j:j+len(sep)] } !(j in indices) && InRangeInc(j, idx, len(s) - len(sep)) && !SetContainsInRange(indices, j-len(sep), j) ==> s[j:j+len(sep)] != sep
+	assume forall j int :: { j in indices }{ s[j:j+len(sep)] } !(j in indices) && InRangeInc(j, 0, len(s) - len(sep)) && !SetContainsInRange(indices, j-len(sep), j) ==> s[j:j+len(sep)] != sep
+}
+
+
+
+// Repeat
+
+
+ghost
+requires bp < lnb
+requires bp == lb * i
+requires lnb == lb * count
+requires 0 <= lb
+ensures  i < count
+decreases
+func lemmaStillHaveToRepeat(i, count, bp, lb, lnb int) {
+	if lb != 0 {
+		assert lb * i < lnb
+		assert lb * i < lb * count
+		LemmaMulInjLtLeftInv(i, count, lb)
+		assert i < count
+	}
+}
+
+
+
+
+
+ghost
+requires InRangeInc(bp, 0, len(nb))
+requires InRangeInc(nCopied, 0, bp)
+requires bp + nCopied <= len(nb)
+requires acc(sl.Bytes(nb, 0, len(nb)), R50)
+requires unfolding acc(sl.Bytes(nb, 0, len(nb)), R50) in forall i int :: {&nb[bp:][i]} 0 <= i && i < len(nb[bp:]) ==> &nb[bp:][i] == &nb[i+bp]
+requires unfolding acc(sl.Bytes(nb, 0, len(nb)), R50) in forall i int :: {&nb[bp:][i]}{&nb[:bp][i]} 0 <= i && i < nCopied ==>  nb[bp:][i] ==  nb[:bp][i]
```

Comment:
Can we use a function here to avoid unfolding in the precondition? If need be, we can introduce a new one